Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

```diff
@@ -0,0 +1,200 @@
+    (Over.mapComp g f).symm
+
+/-- In cartesian monoidal categories, any morphism to the terminal object has a chosen pullback. -/
+@[simps]
+def cartesianMonoidalCategoryToTerminal [CartesianMonoidalCategory C] {X : C} (f : X ⟶ 𝟙_ C) :
+    ChosenPullback f where
+  pullback.obj Y := Over.mk (fst X Y.left)
+  pullback.map {Y Z} g := Over.homMk (X ◁ g.left)
+  mapPullbackAdj := Adjunction.mkOfHomEquiv
+    { homEquiv U Z :=
+      { toFun z := Over.homMk (lift U.hom z.left)
+        invFun u := Over.homMk (u.left ≫ snd X Z.left)
+        left_inv k := by simp
+        right_inv k := by
+          ext
+          dsimp
+          ext
+          · simpa using k.w.symm
+          · aesop } }
+
+/-- In cartesian monoidal categories, the second product projections `snd` have chosen pullbacks. -/
+@[simps]
+def cartesianMonoidalCategorySnd [CartesianMonoidalCategory C] {X Y : C} :
+    ChosenPullback (snd X Y : X ⊗ Y ⟶ Y) where
+  pullback.obj Z := Over.mk (X ◁ Z.hom)
+  pullback.map g := Over.homMk (X ◁ g.left)
+  mapPullbackAdj.unit.app T := Over.homMk (lift (T.hom ≫ fst _ _) (𝟙 _))
+  mapPullbackAdj.counit.app U := Over.homMk (snd _ _)
+
+section PullbackFromChosenPullbacks
+
+variable {Y Z X : C} (f : Y ⟶ X) (g : Z ⟶ X) [ChosenPullback g]
+
+/-- The underlying object of the chosen pullback along `g` of `f`. -/
+abbrev pullbackObj := ((pullback g).obj (Over.mk f)).left
+
+/-- A morphism in `Over X` from the chosen pullback along `g` of `f` to `Over.mk f`. -/
+abbrev fst' := (mapPullbackAdj g).counit.app (Over.mk f)
+
+/-- The first projection from the chosen pullback along `g` of `f` to the domain of `f`. -/
+abbrev fst : pullbackObj f g ⟶ Y := fst' f g |>.left
+
+theorem fst'_left : (fst' f g).left = fst f g := by
+  rfl
+
+/-- The second projection from the chosen pullback along `g` of `f` to the domain of `g`. -/
+abbrev snd : pullbackObj f g ⟶ Z := (pullback g).obj (Over.mk f) |>.hom
+
+/-- A morphism in `Over X` from the chosen pullback along `g` of `f` to `Over.mk g`. -/
+abbrev snd' : (Over.map g).obj ((pullback g).obj (Over.mk f)) ⟶ (Over.mk g) :=
+  Over.homMk (snd f g)
+
+theorem snd'_left : (snd' f g).left = snd f g := by
+  rfl
+
+variable {f g}
+
+@[reassoc]
+theorem condition : fst f g ≫ f = snd f g ≫ g :=
+  Over.w (fst' f g)
+
+variable (f g) in
+
+@[ext]
+theorem hom_ext {W : C} {φ₁ φ₂ : W ⟶ pullbackObj f g} (h₁ : φ₁ ≫ fst _ _ = φ₂ ≫ fst _ _)
+    (h₂ : φ₁ ≫ snd _ _ = φ₂ ≫ snd _ _) :
+    φ₁ = φ₂ := by
+  let adj := mapPullbackAdj g
+  let U : Over Z := Over.mk (φ₁ ≫ snd f g)
+  let φ₁' : U ⟶ (pullback g).obj (Over.mk f) := Over.homMk φ₁
+  let φ₂' : U ⟶ (pullback g).obj (Over.mk f) := Over.homMk φ₂ (by simpa using h₂.symm)
+  have : (adj.homEquiv U _).symm φ₁' = (adj.homEquiv U _).symm φ₂' := by
+    simp [adj.homEquiv_symm_apply]
+    apply (forget X).map_injective
+    simpa using h₁
+  have : φ₁' = φ₂' := by
+    apply (adj.homEquiv U _).symm.injective
+    exact this
```

Comment:
Is it possible to move the proof part of the first `have` in place of the `exact this` in the second `have`?

Reply:
Also, the `simp` is nonterminal.